text/0000-adt-kinds.md

```diff
@@ -0,0 +1,181 @@
+`US{..s}` and in struct patterns `US{}`/`US{..}`. In both cases the path `US` of the expression
+or pattern is looked up in the type namespace (so these expressions/patterns can be used with type
+aliases).
+Fields of a unit struct could also be accessed with dot syntax, but it doesn't have any fields.
+
+As a constant, a unit struct can participate in unit struct expressions `US` and unit struct
+patterns `US`, both of these are looked up in the value namespace in which the constant `US` is
+defined (so these expressions/patterns cannot be used with type aliases).
+
+Note 1: the constant is not exactly a `const` item, there are subtle differences (e.g. with regards
+to `match` exhaustiveness), but it's a close approximation.  
+Note 2: the constant is pretty weirdly namespaced in case of unit *variants*, constants can't be
+defined in "enum modules" manually.
+
+## Tuple structs
+
+Tuple structs are declared with parentheses.
+```
+struct TS(Type0, Type1, Type2);
+```
+
+Tuple structs can be thought of as a single declaration for two things: a basic struct
+
+```
+struct TS {
+    0: Type0,
+    1: Type1,
+    2: Type2,
+}
+```
+
+and a constructor function with the same name<sup>Note 2</sup>
+
+```
+fn TS(arg0: Type0, arg1: Type1, arg2: Type2) -> TS {
+    TS{0: arg0, 1: arg1, 2: arg2}
+}
+```
+
+Tuple structs have 0 or more automatically-named fields and are defined in both type (the type `TS`)
+and the value (the constructor function `TS`) namespaces.
+
+As a basic struct, a tuple struct can participate in struct expressions `TS{0: expr, 1: expr}`,
+including FRU `TS{0: expr, ..ts}`/`TS{..ts}` and in struct patterns
+`TS{0: pat, 1: pat}`/`TS{0: pat, ..}`/`TS{..}`.
+In both cases the path `TS` of the expression or pattern is looked up in the type namespace (so
+these expressions/patterns can be used with type aliases).
+Fields of a tuple struct can be accessed with dot syntax `ts.0`.
+
+As a constructor, a tuple struct can participate in tuple struct expressions `TS(expr, expr)` and
+tuple struct patterns `TS(pat, pat)`/`TS(..)`, both of these are looked up in the value namespace
+in which the constructor `TS` is defined (so these expressions/patterns cannot be used with type
+aliases). Tuple struct expressions `TS(expr, expr)` are usual
+function calls, but the compiler reserves the right to make observable improvements to them based
+on the additional knowledge, that `TS` is a constructor.
+
+Note 1: the automatically assigned field names are quite interesting, they are not identifiers
+lexically (they are integer literals), so such fields can't be defined manually.  
+Note 2: the constructor function is not exactly a `fn` item, there are subtle differences (e.g. with
+regards to privacy checks), but it's a close approximation.
+
+## Summary of the changes.
+
+Everything related to braced structs and unit structs is already implemented.
+
+New: Permit tuple structs and tuple variants with 0 fields. This restriction is artificial and can
```

Comment:
`struct TS();` is roughly equivalent to

```
struct TS {
}
 
 fn TS() -> TS {
     TS{}
 }
```
, it's described in the section about tuple structs above.